src/charm.py

```diff
@@ -53,6 +65,98 @@ def _on_peer_relation_changed(self, event):
             ] = event.relation.data[event.app]["MYSQL_ROOT_PASSWORD"]
             logger.info("Storing MYSQL_ROOT_PASSWORD in StoredState")
 
+    def _on_config_changed(self, _):
+        """This method handles the .on.config_changed() event"""
+        self._configure_pod()
+
+    # Handles start event
+    def _on_start(self, event):
+        """Initialize MySQL
+
+        This event handler is deferred if initialization of MySQL
+        fails. By doing so it is gauranteed that another
+        attempt at initialization will be made.
+        """
+
+        if not self.unit.is_leader():
+            return
+
+        if not self.mysql.is_ready():
+            msg = "Waiting for MySQL Service"
+            self.unit.status = WaitingStatus(msg)
+            logger.debug(msg)
+            event.defer()
+            return
+
+        self._on_update_status(event)
+        self._stored.mysql_initialized = True
+        self.unit.status = ActiveStatus()
+
+    # Handles update-status event
+    def _on_update_status(self, event):
+        """Set status for all units
+        # FIXME
+        Status may be
+        - MySQL is not ready,
+        - MySQL is not Initialized
+        - Unit is active
+        """
+        if not self.unit.is_leader():
+            self.unit.status = ActiveStatus()
+            return
+
+        if not self.mysql.is_ready():
+            status_message = "MySQL not ready yet"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        if not self._stored.mysql_initialized:
+            status_message = "MySQL not initialized"
+            self.unit.status = WaitingStatus(status_message)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    @property
+    def mysql(self) -> MySQL:
+        """Returns MySQL object"""
+        mysql_config = {
+            "app_name": self.model.app.name,
+            "host": self.hostname,
+            "port": self.model.config["port"],
+            "user_name": "root",
+            "mysql_root_password": self._stored.mysql_setup[
+                "MYSQL_ROOT_PASSWORD"
+            ],
+        }
+        return MySQL(mysql_config)
+
+    @property
+    def hostname(self) -> str:
+        """Unit hostname"""
+        unit_id = self.unit.name.split("/")[1]
+        return "{0}-{1}.{0}-endpoints".format(self.model.app.name, unit_id)
+
+    @property
+    def provides(self) -> dict:
+        """Provides dictionary"""
+        provides = {
+            "provides": {"mysql": self.mysql.version()},
+            "config": self.db_info(),
+        }
+        return provides
+
+    # @property
+    def db_info(self) -> dict:
+        info = {
+            "app_name": self.model.app.name,
+            "host": self.hostname,
+            "port": self.model.config["port"],
+            "user_name": "root",
```

Comment:
I know I do this myself. I think we both should add some logic to return relation (client) specific username and passwords. There may be differences here for relational and non relational databases. For example if you have decided to let the client issue the query to create a database then the user created for the client will need to be have create privilege. On the other hand if you decide not to let the client create the data base then you charm lib `new_database` API may need an additional parameter, which is the database schema.

Reply:
I have pushed a prototype of client specific credentials to my [mongodb-operator branch](https://github.com/balbirthomas/mongodb-operator/tree/mongodb-charm-libs).

Reply:
Looks good @balbirthomas I'm going to start implementing something like this